_posts/2024-08-27-building-browser.md

```diff
@@ -0,0 +1,175 @@
+    alt="Verso displaying ASCII texts with CRT style"></a>
+<figcaption>Figure 1: Verso window displaying two different webviews. One for UI, the other for webpage.</figcaption></figure>
+
+For now, the inter-process communication is done via Servo’s existing channel messages like [EmbedderMsg](https://doc.servo.org/embedder_traits/enum.EmbedderMsg.html) and [EmbedderEvent](https://doc.servo.org/compositing/windowing/enum.EmbedderEvent.html).
+We are looking to improve the IPC mechanism with more granular control over DOM elements.
+So, the panel UI can be updated based on the status of web pages.
+One example is when the page URL is changed and the navigation bar needs to be updated.
+There are some candidates for this, such as [WebDriverCommandMsg](https://doc.servo.org/script_traits/enum.WebDriverCommandMsg.html).
+[@webbeef](https://github.com/webbeef) also started a discussion about [defining custom elements](https://github.com/servo/servo/discussions/32883) like `<web-view>` for better ergonomics.
+Overall, improving IPC will be the next target to research after initial multi-view support.
+We will also define more specific webview types to satisfy different purposes in the future.
+
+## Multi-window
+
+The other prominent feature after multi-view is the ability to support multiple windows.
+This one wasn’t planned at first, but because it affects too many components, we ended up resolving them together from the ground up.
+
+Servo uses [WebRender](https://github.com/servo/webrender), based on OpenGL, to render its layout.
+To support multiple windows, we need to support multiple OpenGL surfaces.
+One approach would be to create separate OpenGL contexts for each window.
+But since our implementations of WebGL, WebGPU, and WebXR are all tied to a single WebRender instance, which in turn only supports a single OpenGL context for now, we chose to use a single context with multiple surfaces.
+This alternative approach could potentially use less memory and spawn fewer threads.
+For more details, see [this series of blog posts](https://wusyong.github.io/posts/verso-compositor-part3/) by [@wusyong](https://github.com/wusyong).
+
+<figure class="_fig"><a href="{{ '/img/blog/multiwindow.png' | url }}"><img src="{{ '/img/blog/multiwindow.png' | url }}"
+    alt="Verso displaying two windows"></a>
+<figcaption>Figure 2: Verso creates two different windows with same OpenGL context.</figcaption></figure>
+
+There is still room for improvement.
+For example, WebRender currently [only supports rendering a single “document”](https://github.com/servo/webrender/blob/0.64/webrender/src/renderer/mod.rs#L1228).
+This means this document has to constantly update different display lists to show on other windows, which may cause race conditions and let unintended webview display on the wrong window in a slip second.
+<!-- why is this not a problem in Firefox? -->
```

Comment:
Firefox create [another webrender](https://searchfox.org/mozilla-central/source/gfx/webrender_bindings/src/bindings.rs#1781) whenever a new window is created.
There's a [TODO](https://github.com/servo/webrender/blob/main/webrender/src/renderer/mod.rs#L1016) in webrender to support multiple documents rendering. I would like to explore this way to see if we can just use single webrender to support multiple window.

Reply:
reworked to include this in 6c730ac5c5e1284713b9de6aa51377211d15f5e0